Src/StdLib/Lib/_weakrefset.py

```diff
@@ -70,9 +70,9 @@ def __len__(self):
     def __contains__(self, item):
         try:
             wr = ref(item)
+            return wr in self.data
```

Comment:
I'm not saying there's no issue here, but I'm not really sure I follow the reasoning for this change. What exactly is getting GCed here and how does this fix the issue? Looking at the `abc` module it seems to be maintaining the reference to `item` so it shouldn't be collected. And the weakref object certainly isn't getting collected before the end of the operations.

Reply:
Yes - it does seem odd that having the item as a parameter is not forcing that item in memory.

I'm trying to think of other possible ways you could get the error from that line...

What about if one of the other items in the set has been freed - perhaps CPython is just testing the weakref provided against the pointer location of the weakrefs in the set to establish equality whereas IronPython might be dereferencing each item in the set and comparing the inner items? That would cause the symptom I'm seeing... I guess it would mean the fix should be in the equality check tho rather than what I proposed.

I'll see if I can debug into the comparison and see what its doing

Reply:
So I spent today playing with this and, while I dont really understand why, my original diagnosis appears to be correct.

The exception comes from the __hash__ function of weakref.ref at _weakref.cs line 219 when the "in" operator is called. So either the items in the set have been freed and their hashes fail or the item in wr has been freed and its hash fails.

You can run the following code to explore the behaviour of the weakref.ref hash function and get the exception:

```
class A(): pass

# first show that weakref.ref hash works normally
a = A()
import weakref
r = weakref.ref(a)
print hash(r)
# prints hash

# now show that the hash is cached
del a
import gc
gc.collect(2)
print hash(r)
# prints same hash

# now show that, if you dont calculate the hash before you dispose the item then it fails
del r
a = A()
r = weakref.ref(a)
del a
gc.collect(2)
print hash(r)
# error
```

So, looking at the implementation of Set, you can see that it caches the hashes as soon as you add an item to the list. This suggests that its "item" that has been freed, not other items in the set.

To confirm this, you can undo my original fix and instead force the cache of the hash:

```
    def __contains__(self, item):
        try:
            wr = ref(item)
            hash(wr)
        except TypeError:
            return False
        return wr in self.data
```

Running the original test 100 times shows this also fixes the issue.

From other testing I've done, I've seen that try/except blocks seem to sometimes (but not reliably?) form scope boundaries for IronPython so my suspicion is that item is being "released" when leaving the try/except block and in the 3 out of 100 case the GC fires just at the wrong time and blows item away (I'm not sure why, as you say, you'd expect the stack frame entering __contains__ to be holding a reference to item). I cant reproduce this by manually calling the gc so I feel I'm on shaky ground with this assertion. 

Have a look at the attached testscope script for various tests around scoping and gc - there's some oddities and I did manage to sort of recreate the bug with slightly modified code where "item" is itself a weakref...
[testscope.zip](https://github.com/IronLanguages/ironpython2/files/1498995/testscope.zip)

Reply:
BTW - I also ran the original bug script against CPython and cannot reproduce this TypeError there even tho I can see the ipy and python code are pretty much line-for-line ports all the way down. Some aspect of CPython's threading / GC must be happily blocking this issue from occurring there.

Reply:
Thanks for your analysis. I concur that the only way the exception would occur is if `item` has been GCed before the hashing occurs. I just don't get how it happens since I would expect `item` to be held alive until the end of the function call (as your `test3` shows).

I'm fine with the proposed change to `_weakrefset.py `, however you should mark it up with a comment referencing the issue or PR otherwise it may get reverted when we update the standard library to sync up with CPython.

As for the testscope script here are my comments:
- `test3` is behaving as expected in that argument should be held alive until the end of the function call.
- `test4` only being cleaned up in `test5` is certainly unexpected - seems like a bug
- `test6` is also unexpected and seems like a bug
- `test7` is behaving as expected, with IPy there should be two possible outcomes:
  - GC occurs between the assignment to `a` and the call to `test7` in which case the value of `b` is `None` which gives the same result as CPython
  - Otherwise the GC hasn't occurred and the value of `b` is `A(7)`. In this case `b` should be held alive until the function returns
- `test8` is also behaving as expected. Like test7 there are two possible outcomes. The value of `b` is always the weakref but its target has either been collected of not:
  - GC occurs between the assignment to `a` and the call to `test8` in which case the target of `b` is `None` and we get the same result as CPython.
  - Otherwise the GC hasn't occurred and the target of `b` is still `A(8)`. However in this case since `b` is a weakref to `A(8)` the GC can collect still clean up `A(8)`.

Reply:
Added one more commit to comment the change in _weakrefset.py

Thanks for looking at testscope - I agree with your comments. Should I add issues to cover those test cases?

Reply:
Sure, better off adding an issue to cover those cases otherwise I'm likely to forget about it.